File: src/opsiq_runtime/app/api/models/decisions.py
"""Pydantic models for decision-related API responses."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class DecisionKey(BaseModel):
    """Decision key identifying a unique decision."""

    tenant_id: str
    subject_type: str
    subject_id: str
    primitive_name: str
    primitive_version: str
    as_of_ts: datetime = Field(..., description="ISO timestamp")


class DecisionListItem(DecisionKey):
    """Decision item for list responses."""

    decision_state: str = Field(..., description="URGENT/WATCHLIST/HEALTHY/UNKNOWN")
    confidence: str = Field(..., description="HIGH/MEDIUM/LOW")
    computed_at: datetime = Field(..., description="ISO timestamp")
    drivers: list[str] = Field(default_factory=list, description="Parsed JSON array")
    metrics: dict[str, Any] = Field(default_factory=dict, description="Parsed JSON object")


class DecisionListResponse(BaseModel):
    """Response for decision list endpoints."""

    items: list[DecisionListItem]
    next_cursor: str | None = None


class DecisionDetail(DecisionKey):
    """Detailed decision information."""

    canonical_version: str
    config_version: str
    decision_state: str = Field(..., description="URGENT/WATCHLIST/HEALTHY/UNKNOWN")
    confidence: str = Field(..., description="HIGH/MEDIUM/LOW")
    computed_at: datetime = Field(..., description="ISO timestamp")
    valid_until: datetime | None = None
    drivers: list[str] = Field(default_factory=list, description="Parsed JSON array")
    metrics: dict[str, Any] = Field(default_factory=dict, description="Parsed JSON object")
    evidence_refs: list[str] = Field(default_factory=list, description="Parsed JSON array of evidence IDs")
    correlation_id: str | None = None


class EvidenceRecord(BaseModel):
    """Evidence record from the evidence table."""

    tenant_id: str
    evidence_id: str
    primitive_name: str
    primitive_version: str
    as_of_ts: datetime = Field(..., description="ISO timestamp")
    computed_at: datetime = Field(..., description="ISO timestamp")
    evidence: dict[str, Any] = Field(default_factory=dict, description="Parsed JSON object")


class DecisionBundle(BaseModel):
    """Decision bundle with composite, components, and evidence."""

    composite: DecisionDetail
    components: dict[str, DecisionDetail] = Field(
        default_factory=dict,
        description="Component decisions keyed by primitive_name (operational_risk, shopper_frequency_trend)",
    )
    evidence: dict[str, list[EvidenceRecord]] = Field(
        default_factory=dict,
        description="Evidence grouped by primitive_name (composite, operational_risk, shopper_frequency_trend)",
    )

File: frontend/types/decisions.ts
/**
 * TypeScript types matching the FastAPI backend Pydantic models.
 */

export interface DecisionKey {
  tenant_id: string
  subject_type: string
  subject_id: string
  primitive_name: string
  primitive_version: string
  as_of_ts: string // ISO timestamp
}

export interface DecisionListItem extends DecisionKey {
  decision_state: "URGENT" | "WATCHLIST" | "HEALTHY" | "UNKNOWN"
  confidence: "HIGH" | "MEDIUM" | "LOW"
  computed_at: string // ISO timestamp
  drivers: string[]
  metrics: Record<string, any>
}

export interface DecisionListResponse {
  items: DecisionListItem[]
  next_cursor: string | null
}

export interface DecisionDetail extends DecisionKey {
  canonical_version: string
  config_version: string
  decision_state: "URGENT" | "WATCHLIST" | "HEALTHY" | "UNKNOWN"
  confidence: "HIGH" | "MEDIUM" | "LOW"
  computed_at: string // ISO timestamp
  valid_until: string | null
  drivers: string[]
  metrics: Record<string, any>
  evidence_refs: string[]
  correlation_id: string | null
}

export interface EvidenceRecord {
  tenant_id: string
  evidence_id: string
  primitive_name: string
  primitive_version: string
  as_of_ts: string // ISO timestamp
  computed_at: string // ISO timestamp
  evidence: Record<string, any>
}

export interface DecisionBundle {
  composite: DecisionDetail
  components: Record<string, DecisionDetail> // keyed by primitive_name
  evidence: Record<string, EvidenceRecord[]> // grouped by primitive_name
}

export interface DecisionHistoryItem {
  primitive_name: string
  primitive_version: string
  as_of_ts: string // ISO timestamp
  decision_state: string
  confidence: "HIGH" | "MEDIUM" | "LOW"
  drivers: { code: string }[]
  computed_at: string // ISO timestamp
}

export interface DecisionHistoryResponse {
  subject_id: string
  items: DecisionHistoryItem[]
}

